Sources/StreamVideo/WebRTC/v2/PeerConnection/Adapters/ICEAdapter.swift

```diff
@@ -0,0 +1,278 @@
+//
+// Copyright © 2024 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+import StreamWebRTC
+
+actor ICEAdapter: @unchecked Sendable {
+    private let sessionID: String
+    private let encoder = JSONEncoder()
+    private let peerType: PeerConnectionType
+    private let peerConnection: RTCPeerConnection
+
+    private var disposableBag: DisposableBag = .init()
+    private var trickledCandidates: [RTCIceCandidate] = []
+    private var untrickledCandidates: [RTCIceCandidate] = []
+
+    var sfuAdapter: SFUAdapter
+
+    init(
+        sessionID: String,
+        peerType: PeerConnectionType,
+        peerConnection: RTCPeerConnection,
+        sfuAdapter: SFUAdapter
+    ) {
+        self.sessionID = sessionID
+        self.peerType = peerType
+        self.peerConnection = peerConnection
+        self.sfuAdapter = sfuAdapter
+
+        Task {
+            await configure()
+        }
+    }
+
+    func trickle(_ candidate: RTCIceCandidate) {
+        guard case .connected = sfuAdapter.connectionState else {
+            untrickledCandidates.append(candidate)
+            return
+        }
+        trickleTask(for: candidate)
+            .store(in: disposableBag)
+    }
+
+    private func trickleTask(
+        for candidate: RTCIceCandidate
+    ) -> Task<Void, Never> {
+        Task {
```

Comment:
I don't see any issues. The task will execute up to the point where we go to the SFUAdapter. At this point the actor will pick up something else to work on until the SFUAdapter returns. When we get the result from the SFUAdapter the actor will atomically update the trickedCandidates by only appending (not removing that's an important detail). Meaning that even in concurrent tasks trying to modify the array things will be as expected (indexes only increasing and we are relying on appending rather than pure indexes, so count doesn't matter).